lighthouse-core/audits/seo/font-size.js

```diff
@@ -0,0 +1,233 @@
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions and limitations under the License.
+ */
+'use strict';
+
+const parseURL = require('url').parse;
+const Audit = require('../audit');
+const ViewportAudit = require('../viewport');
+const CSSStyleDeclaration = require('../../lib/web-inspector').CSSStyleDeclaration;
+const MINIMAL_PERCENTAGE_OF_LEGIBLE_TEXT = 75;
+
+/**
+ * @param {Array<{cssRule: WebInspector.CSSStyleDeclaration, fontSize: number, textLength: number, node: Node}>} fontSizeArtifact
+ * @returns {Array<{cssRule: WebInspector.CSSStyleDeclaration, fontSize: number, textLength: number, node: Node}>}
+ */
+function getUniqueFailingRules(fontSizeArtifact) {
+  const failingRules = new Map();
+
+  fontSizeArtifact.forEach(({cssRule, fontSize, textLength, node}) => {
+    const artifactId = getFontArtifactId(cssRule, node);
+    const failingRule = failingRules.get(artifactId);
+
+    if (!failingRule) {
+      failingRules.set(artifactId, {
+        node,
+        cssRule,
+        fontSize,
+        textLength,
+      });
+    } else {
+      failingRule.textLength += textLength;
+    }
+  });
+
+  return failingRules.valuesArray();
+}
+
+/**
+ * @param {Node} node
+ * @return {{type:string, snippet:string}}
+ */
+function nodeToTableNode(node) {
+  const attributesString = node.attributes.map((value, idx) =>
+    (idx % 2 === 0) ? ` ${value}` : `="${value}"`
+  ).join('');
+
+  return {
+    type: 'node',
+    snippet: `<${node.localName}${attributesString}>`,
+  };
+}
+
+/**
+ * @param {string} baseURL
+ * @param {WebInspector.CSSStyleDeclaration} styleDeclaration
+ * @param {Node} node
+ * @returns {{source:!string, selector:string|object}}
+ */
+function findStyleRuleSource(baseURL, styleDeclaration, node) {
+  if (
+    !styleDeclaration ||
+    styleDeclaration.type === CSSStyleDeclaration.Type.Attributes ||
+    styleDeclaration.type === CSSStyleDeclaration.Type.Inline
+  ) {
+    return {
+      source: baseURL,
+      selector: nodeToTableNode(node),
+    };
+  }
+
+  if (styleDeclaration.parentRule &&
+    styleDeclaration.parentRule.origin === global.CSSAgent.StyleSheetOrigin.USER_AGENT) {
+    return {
+      selector: styleDeclaration.parentRule.selectors.map(item => item.text).join(', '),
+      source: 'User Agent Stylesheet',
+    };
+  }
+
+  if (styleDeclaration.type === CSSStyleDeclaration.Type.Regular && styleDeclaration.parentRule) {
+    const rule = styleDeclaration.parentRule;
+    const stylesheet = styleDeclaration.stylesheet;
+
+    if (stylesheet) {
+      const url = parseURL(stylesheet.sourceURL, baseURL);
+      const range = styleDeclaration.range;
+      const selector = rule.selectors.map(item => item.text).join(', ');
+      let source = `${url.href}`;
```

Comment:
there seem to be quite a few where the `url.href` isn't set, does that mean they came from the root document? maybe we can fill in with "<index>" or something instead of empty? it threw me off a bit

![image](https://user-images.githubusercontent.com/2301202/33564256-ef70b8ec-d8ce-11e7-8a7a-bcb08c3534b6.png)

Reply:
>  fill in with ""

🤔 

Normally, source path is shown before line and column number, but when CSS is dynamically injected its source URL is not set:
<img width="659" alt="dynamic-css" src="https://user-images.githubusercontent.com/985504/33638290-affa1598-da25-11e7-891d-bab38cb841f2.png"> .

We could default to the main resource path, but line and column numbers are relative to the injected stylesheet, not the document so it might be confusing. IMO it would make most sense to just say 'dynamically injected' in this case (and drop line and column numbers). WDYT?

Reply:
Hm yeah rather tricky

"dynamically injected" or even just "dynamic" sounds good to me

Reply:
Updated:

<img width="750" alt="dyanmic" src="https://user-images.githubusercontent.com/985504/33671717-aa59ef4a-daa8-11e7-89b7-ff4546841150.png">